documentation/Semantics.md

```diff
@@ -1,206 +1,152 @@
-- Others….? TBD
-
-Resolution of parse tree ambiguity (statement function definition, function vs.
-  array)
-
-#### Step 7. Do enough semantic processing to generate .mod files
-- Fully resolve derived types
-- Combine and check declarations of all entities within a given scope; resolve
-their type, rank, shape, and other attributes.
-- Constant evaluation is required at this point.
-
-Why do Step 7 before the rest of semantic checking? The sooner we can generate
-mod file the sooner we can read ‘em; you can test a lot of Fortran programs as
-soon as you can read mod files.
-
-#### Step 8. Semantic Rule Checking
-
-An incomplete and unordered list of requirements for semantic analysis:
-
-* EQUIVALENCE overlaying (checking at least)
-* Intrinsic function generic->specific resolution, constraint checking, T/R/S.
-* Compile-time evaluation of constant expressions, including intrinsic
-functions.
-* Resolution of generics and type-bound procedures.
-* Identifying and recording uplevel references.
-* Control flow constraint checking
-* Labeled DO loop terminal statement expansion? (maybe not, can defer to CFG in
-  DST).
-* Construct association: distinguish pointer-like from allocatable-like
-* OMP and OACC checking
-* CUF constraint checking
-
-## Utility Routines
-
-### Diagnostic Output
-TBD
-
-### Constant Expression Evaluation
-- Scalars
-- Array intrinsics
+If the program is legal, the semantic pass will produce a (possibly modified)
+parse tree for the semantically correct program with each name mapped to a symbol
+and each expression fully analyzed.
+
+All user errors are detected either prior to or during semantic analysis.
+After it completes successfully the program should compile with no error messages.
+There may still be warnings or informational messages.
+
+## Phases of Semantic Analysis
+
+1. [Validate labels](#validate-labels) -
+   Check all constraints on labels and branches
+2. [Rewrite DO loops](#rewrite-do-loops) -
+   Convert all occurrences of `LabelDoStmt` to `DoConstruct`.
+3. [Name resolution](#name-resolution) -
+   Analyze names and declarations, build a tree of Scopes containing Symbols,
+   and fill in the `Name::symbol` data member in the parse tree
+4. [Rewrite parse tree](#rewrite-parse-tree) -
+   Fix incorrect parses based on symbol information
+5. [Expression analysis](#expression-analysis) -
+   Analyze all expressions in the parse tree and fill in `Expr::typedExpr` and
+   `Variable::typedExpr` with analyzed expressions; fix incorrect parses
+   based on the result of this analysis
+6. [Statement semantics](#statement-semantics) -
+   Perform remaining semantic checks on the execution parts of subprograms
+7. [Write module files](#write-module-files) -
+   If no errors have occurred, write out `.mod` files for modules and submodules
+
+If phase 1 or phase 2 encounter an error on any of the program units,
+compilation terminates. Otherwise, phases 3-6 are all performed even if
+errors occur.
```

Comment:
We need to think about whether this is the right thing to do. Maybe we can keep going if some of the branch and label constraints aren't met. But if we are unable to convert a label DO to a DO construct, things may be too messed up to proceed.

Also, what if there are multiple program units and one has an error? Should we continue with the others? If the one with errors is a module, other program units that USE it probably shouldn't be analyzed further.